### PR TITLE
Don't apply default search filters, aggregators or matchers in search tests

### PR DESCRIPTION
**Pull request by hmstepanek**
Make the default search filters, aggregators, and matchers clearable so that we can remove them when unit testing so as to only test one filter at a time.

### CODE REVIEW

File: h/search/core.py [hunk 1]
```diff
@@ -16,6 +16,19 @@
...
     'aggregations'])
 
+DEFAULT_FILTERS = [
+    query.DeletedFilter,
+    query.GroupFilter,
+    query.UserFilter,
+    query.AnyMatcher,
+    query.TagsMatcher]
+
+DEFAULT_REQUEST_FILTERS = [
```
Comment (hmstepanek) [inline h/search/core.py]: There are currently two types of filters, ones that take a request obj as an arg and ones that don't hence the two different lists here.

File: h/search/core.py [hunk 2]
```diff
@@ -145,15 +178,10 @@ def _instrument(self):
             s.send()
 
     @staticmethod
-    def _default_querybuilder(request):
+    def _default_querybuilder(request, filters, request_filters):
```
Comment (hmstepanek) [inline h/search/core.py]: We could split this into two separate methods; one that adds reg filters and one that adds request filters. Not sure if that's necessary or not??

File: tests/h/search/query_test.py [hunk 3]
```diff
@@ -3,188 +3,182 @@
...
 
 class TestAuthorityFilter(object):
-    def test_it_filters_out_non_matching_authorities(self, Annotation, search):
-        annotations_auth1 = [Annotation(userid="acct:foo@auth1", shared=True).id,
-                             Annotation(userid="acct:bar@auth1", shared=True).id]
+    def test_it_filters_out_non_matching_authorities(self, Annotation, filtered_search):
+        annotations_auth1 = [Annotation(userid="acct:foo@auth1").id,
+                             Annotation(userid="acct:bar@auth1").id]
         # Make some other annotations that are of different authority.
-        Annotation(userid="acct:bat@auth2", shared=True)
```
Comment (hmstepanek) [inline tests/h/search/query_test.py]: Since we aren't testing with the shared filter anymore we don't need to mark all these as shared annotations.

File: h/search/core.py (hunk 1 above)
Comment (robertknight) [reply h/search/core.py]: Having two separate lists seems a perfectly pragmatic approach to me. I would suggest adding a comment above each variable declaration to just clarify the difference.

File: h/search/core.py (hunk 2 above)
Comment (robertknight) [reply h/search/core.py]: This looks fine to me.

File: tests/h/search/query_test.py (hunk 3 above)
Comment (robertknight) [reply tests/h/search/query_test.py]: 👍

File: tests/h/search/query_test.py [hunk 4]
```diff
@@ -3,188 +3,182 @@
...
 
 class TestGroupFilter(object):
 
     @pytest.fixture
-    def search(self, pyramid_request):
-        # We don't need to append GroupFilter to Search because it's one of the filters that
-        # Search appends by default.
-        return search.Search(pyramid_request)
+    def filtered_search(self, search):
+        search.append_aggregation(query.GroupFilter())
```
Comment (robertknight) [inline tests/h/search/query_test.py]: Why is this `append_aggregation` and not `append_filter`?

Comment (hmstepanek) [reply tests/h/search/query_test.py]: Me too...lol I thought I removed all of those but I guess I didn't. I'll fix that.

**Comment by hmstepanek** [comment]
Ok changes addressed; added comments above default filter lists and now using append_filter everywhere.

**Comment by hmstepanek** [comment]
Changes addressed: using append_matcher/append_filter and added comments above default filter lists.

File: tests/h/search/query_test.py [hunk 5]
```diff
@@ -3,188 +3,182 @@
...
 
 class TestTagsAggregation(object):
 
     @pytest.fixture
-    def search(self, pyramid_request):
-        search_ = search.Search(pyramid_request)
-        search_.append_aggregation(search.TagsAggregation())
-        return search_
+    def filtered_search(self, search):
+        search.append_filter(query.TagsAggregation())
```
Comment (robertknight) [inline tests/h/search/query_test.py]: This should be `append_aggregation` surely? Same with `TestUsersAggregation` below.

**Comment by hmstepanek** [comment]
Yeah sorry...just noticed another issue too. Durp.

**Comment by seanh** [comment]
Can we remove <https://github.com/hypothesis/h/pull/5066/commits/9e825415f356a8700154489a35bf422414e4efbe> (_Remove shared=True from search filter test cases_)? I don't think we should do that - I think the tests should be working with shared annotations, because that's the kind of data that we mostly have in production. Private annotations are a corner case. I wouldn't want all of our search tests to test only with private annotations.

**Comment by seanh** [comment]
Hannah talked to me about this last week and suggested an alternative approach: add a `clear()` method to `Search` that clears all the filters and matchers. That sounds much simpler and less likely to break things (it merely involves _adding_ a tests-only method to `Search`, not refactoring `Search` code that the production code is using), and is also much more elegant than what I was going to suggest. @hmstepanek I wonder if it'd be worth closing this PR, and replacing it with a clean PR for `clear()`?

**Comment by hmstepanek** [comment]
Meh, I’ll just rebase it and change the name. Nbd.

**Comment by seanh** [comment]
Thanks

**Comment by robertknight** [comment]
> Hannah talked to me about this last week and suggested an alternative approach: add a clear() method to Search that clears all the filters and matchers. 

I would probably pick a different name as `clear` is a bit ambiguous (clear what?). My personal stylistic preference is for the approach taken here, which is to initialize objects to the desired state instead of manipulating them after-the-fact. In more complex scenarios, that can eliminate certain kinds of bug (eg. failing to invalidate a cache after modifying object state). Frankly though I think the debate between the two approaches is somewhat philosophical in this context. Either would be fine with me.

**Comment by seanh** [comment]
Remember, this feature is _only needed by the tests_. I think Hannah's suggestion of adding a `clear()` method is going to produce a much smaller and simpler diff - we're talking about adding one roughly two-line method to the class rather than an approximately 50-line refactoring diff, and the method is only called by the tests so it's much less likely to break anything. Given that the current code does in fact break the search code, and activity pages are crashing, I don't think this is philosophical at all.

If either is fine with you, I have a strong preference for seeing how Hannah's clear method idea works out

File: h/search/core.py [hunk 6]
```diff
@@ -58,6 +58,11 @@ def run(self, params):
 
         return SearchResult(total, annotation_ids, reply_ids, aggregations)
 
+    def clear(self):
```
Comment (hmstepanek) [inline h/search/core.py]: I'm open to alternative name suggestions-I would just say clear_filters but technically it clears filters, matchers, and aggregators but to say all that would result in a very long method name. Maybe clear_query?

File: tests/h/search/query_test.py [hunk 7]
```diff
@@ -3,188 +3,182 @@
...
 import pytest
 
-from h import search
+from h.search import Search, query
 
 
 class TestTopLevelAnnotationsFilter(object):
 
-    def test_it_filters_out_replies_but_leaves_annotations_in(self, Annotation, search):
-        annotation = Annotation(shared=True)
```
Comment (hmstepanek) [inline tests/h/search/query_test.py]: I know @seanh mentioned something about keeping the shared=True here. For me it's a debate between writing tests that make sense vs writing true unit tests. I would prefer to keep the shared=True out of these unit tests and move them to the integration tests (aka tests that test with multiple filters/aggregators/matchers). Alternatively, we could keep the shared=True here but then I think we should append the shared filter.

**Comment by hmstepanek** [comment]
I think we can ignore code cov here as it's complaining about the fixtures not being called/covered in the tests module for the unimplemented tests.

File: tests/h/search/query_test.py (hunk 7 above)
Comment (hmstepanek) [reply tests/h/search/query_test.py]: We discussed this and decided to just leave it as is without the shared=True.

File: tests/h/search/query_test.py [hunk 8]
```diff
@@ -3,188 +3,182 @@
...
+        reply = Annotation(references=[annotation.id])
 
-        result = search.run({})
+        result = filtered_search.run({})
 
         assert annotation.id in result.annotation_ids
         assert reply.id not in result.annotation_ids
 
     @pytest.fixture
-    def search(self, pyramid_request):
```
Comment (robertknight) [inline tests/h/search/query_test.py]: Any particular reason for changing the name of the fixture? Usually when we customize a fixture for a particular test case we just override it with one that has the same name.

File: h/search/core.py (hunk 6 above)
Comment (seanh) [reply h/search/core.py]: I think `clear()` is a good name - `search.clear()` is clear enough

File: h/search/core.py [hunk 9]
```diff
@@ -58,6 +58,11 @@ def run(self, params):
...
+        """Clear search filters, aggregators, and matchers."""
+        self.builder = query.Builder()
+        self.reply_builder = query.Builder()
```
Comment (seanh) [inline h/search/core.py]: This is nice - I had been thinking that, since the filters, aggregators and matchers are actually stored and used in `Builder` not in `Search`, you would have to add a `clear()` method to `Builder` and add a `clear()` method to `Search` that calls it. And then you'd be implicitly relying on the fact that `Builder` doesn't actually do anything with an appended filter, matcher or aggregator until `build()` is called so that it's safe to clear them. So just replacing the entire `Builder` objects is much better :+1:

File: h/search/core.py (hunk 6 above)
Comment (seanh) [reply h/search/core.py]: Also the docstring clarifies it perfectly

**Comment by seanh** [comment]
@hmstepanek Can I leave it to you to merge and deploy this? There's a bunch of other PRs queued up on Jenkins that need to be deployed, tested on QA (if they're not just tests PRs), then deployed to production. They all look like PRs that you'll be familiar with

File: tests/h/search/query_test.py (hunk 8 above)
Comment (hmstepanek) [reply tests/h/search/query_test.py]: No particular reason-I'll change it to just override the name I supposed.
